app/src/lib/stores/helpers/infer-comparison-branch.ts

```diff
@@ -0,0 +1,143 @@
+import { Branch, IAheadBehind } from '../../../models/branch'
+import { PullRequest } from '../../../models/pull-request'
+import { Repository } from '../../../models/repository'
+import { GitHubRepository } from '../../../models/github-repository'
+import { revRange } from '../../git'
+
+/**
+ * Infers which branch to use as the comparison branch
+ *
+ * The branch returned is determined by the following conditions:
+ * 1. Given a pull request -> target branch of PR
+ * 2. Given a forked repository -> default branch on `upstream`
+ * 3. Given a hosted repository -> default branch on `origin`
+ * 4. Fallback -> `master` branch
+ *
+ * @param branches The list of all branches for the repository
+ * @param repository The repository the branch belongs to
+ * @param currentPullRequest The pull request to use for finding the branch
+ * @param currentBranch The branch we want the parent of
+ * @param getAheadBehind function used to calculate ahead/behind
+ */
+export async function inferComparisonBranch(
+  branches: ReadonlyArray<Branch>,
+  repository: Repository,
+  currentPullRequest: PullRequest | null,
+  currentBranch: Branch | null,
+  getAheadBehind: (
+    repository: Repository,
+    range: string
+  ) => Promise<IAheadBehind | null>
+): Promise<Branch | null> {
+  if (currentPullRequest !== null) {
+    return _getTargetBranchOfPullRequest(branches, currentPullRequest)
+  }
+
+  const ghRepo = repository.gitHubRepository
+  if (ghRepo !== null) {
+    return ghRepo.fork === true && currentBranch !== null
+      ? _getDefaultBranchOfFork(
+          branches,
+          repository,
+          ghRepo,
+          currentBranch,
+          getAheadBehind
+        )
+      : _getDefaultBranchOfGithubRepo(branches, ghRepo)
+  }
+
+  return _getMasterBranch(branches)
+}
+
+/**
+ * For `inferComparisonBranch` case where inferring for local branches
+ *
+ * Returns `master`, the default branch of a Git repository
+ *
+ * @param branches The list of all branches for the repository
+ */
+function _getMasterBranch(branches: ReadonlyArray<Branch>): Branch | null {
+  return branches.find(b => b.name === 'master') || null
+}
+
+/**
+ * For `inferComparisonBranch` case where inferring for a repository
+ * hosted on GitHub, but with no forks
+ *
+ * Returns the default branch of the GitHub repository
+ *
+ * @param branches The list of all branches for the repository
+ * @param ghRepository The repository the branch belongs to
+ */
+function _getDefaultBranchOfGithubRepo(
```

Comment:
should we capitalize the H in GitHub?

Reply:
I don't believe so since GitHub is one word and not two. But, I have no strong feelings either way, so I'm 👍 to change it.